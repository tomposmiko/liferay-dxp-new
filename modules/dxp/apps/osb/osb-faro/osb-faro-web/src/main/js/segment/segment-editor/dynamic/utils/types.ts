import React from 'react';
import {
	Conjunctions,
	CustomFunctionOperators,
	FunctionalOperators,
	NotOperators,
	RelationalOperators
} from './constants';
import {CustomValue, Property} from 'shared/util/records';

export type Context = {
	lastNodeWasGroup?: boolean;
	oDataASTNode: ODataASTNode;
	prevConjunction?: string;
};

export type CriterionGroup = {
	conjunctionName: string;
	criteriaGroupId: string;
	items?: (CriterionGroup | Criterion)[];
};

export type Criterion = {
	defaultValue?: string;
	operatorName?: Conjunctions &
		CustomFunctionOperators &
		FunctionalOperators &
		NotOperators &
		RelationalOperators;
	propertyName?: string;
	rowId?: string;
	touched?: boolean | object;
	type?: string;
	valid?: boolean | object;
	value?: any;
};

export type Criteria = Criterion | CriterionGroup;

export type OnCriterionAdd = (index: number, criterion: Criterion) => void;

export type ODataASTNode = {
	next?: number;
	position?: number;
	raw?: string;
	type: string;
	value: any;
};

export type OnMove = (
	startGroupId: string,
	startIndex: number,
	destGroupId: string,
	destIndex: number,
	criterion: Criterion | CriterionGroup,
	replace?: boolean
) => void;

export type Operator = {
	key: string;
	label: string;
	name: string;
};

export interface ISegmentEditorInputBase {
	channelId?: string;
	className?: string;
	displayValue?: string;
	groupId?: string;
	id?: string;
	onChange: (params: Criterion | Criterion[]) => void;
	operatorRenderer?: React.ElementType;
	property?: Property;
	timeZoneId?: string;
	touched?: boolean | object;
	valid?: boolean | object;
	value: string | number | CustomValue;
}

export interface ISegmentEditorCustomInputBase extends ISegmentEditorInputBase {
	value: CustomValue;
}
