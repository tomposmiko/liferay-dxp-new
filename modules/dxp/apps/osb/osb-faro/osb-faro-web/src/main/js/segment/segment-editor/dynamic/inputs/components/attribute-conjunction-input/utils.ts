import {BetweenNumber} from '../BetweenNumberInput';
import {
	BOOLEAN_LABELS_MAP,
	DATE_OPERATOR_LONGHAND_LABELS_MAP,
	DATE_OPTIONS,
	DURATION_OPERATOR_LONGHAND_LABELS_MAP,
	DURATION_OPTIONS,
	NUMBER_OPERATOR_LONGHAND_LABELS_MAP,
	NUMBER_OPTIONS
} from 'event-analysis/utils/utils';
import {DataTypes} from 'event-analysis/utils/types';
import {DateRange} from 'shared/components/DateRangeInput';
import {
	FunctionalOperators,
	RelationalOperators,
	STRING_OPERATOR_LABELS_MAP,
	STRING_OPTIONS
} from '../../../utils/constants';
import {isNumber} from 'lodash';
import {isValid} from '../../../utils/utils';

export const createOption = (option, dataType: DataTypes) => {
	const LABELS_MAP = {
		[DataTypes.Boolean]: BOOLEAN_LABELS_MAP,
		[DataTypes.Date]: DATE_OPERATOR_LONGHAND_LABELS_MAP,
		[DataTypes.Duration]: DURATION_OPERATOR_LONGHAND_LABELS_MAP,
		[DataTypes.Number]: NUMBER_OPERATOR_LONGHAND_LABELS_MAP,
		[DataTypes.String]: STRING_OPERATOR_LABELS_MAP // STRING_OPERATOR_LABELS_MAP is provided from the segment-editor utils as "NotContains" differs from segment-editor and event-analysis. We should be able to use the evente-analysis version once we move away from odata.
	};

	return {
		label: LABELS_MAP[dataType][option],
		value: option
	};
};

export const getOperatorOptions = (dataType: DataTypes) => {
	const OPERATOR_OPTIONS = {
		[DataTypes.Date]: DATE_OPTIONS,
		[DataTypes.Duration]: DURATION_OPTIONS,
		[DataTypes.Number]: NUMBER_OPTIONS,
		[DataTypes.String]: STRING_OPTIONS // STRING_OPTIONS is provided from the segment-editor utils as "NotContains" differs from segment-editor and event-analysis. We should be able to use the evente-analysis version once we move away from odata.
	};

	return OPERATOR_OPTIONS[dataType]?.map(option =>
		createOption(option, dataType)
	);
};

export const getDefaultAttributeOperator = (
	dataType: DataTypes
): RelationalOperators | FunctionalOperators => {
	switch (dataType) {
		case DataTypes.Boolean:
		case DataTypes.Date:
			return RelationalOperators.EQ;
		case DataTypes.Duration:
		case DataTypes.Number:
			return RelationalOperators.GT;
		case DataTypes.String:
		default:
			return FunctionalOperators.Contains;
	}
};

export const getDefaultAttributeValue = (
	dataType: DataTypes,
	operatorName: RelationalOperators | FunctionalOperators
): string | {end: number | string; start: number | string} => {
	if (
		operatorName === FunctionalOperators.Between &&
		[DataTypes.Number, DataTypes.Date].filter(type => type === dataType)
	) {
		return {end: '', start: ''};
	}

	switch (dataType) {
		case DataTypes.Boolean:
			return 'true';
		case DataTypes.Date:
		case DataTypes.Number:
		case DataTypes.Duration:
		case DataTypes.String:
		default:
			return '';
	}
};

export const validateAttributeValue = (
	value: string | number | BetweenNumber | DateRange,
	dataType: DataTypes,
	operatorName?: FunctionalOperators | RelationalOperators
): boolean => {
	if (
		operatorName === FunctionalOperators.Between &&
		[DataTypes.Number, DataTypes.Date].filter(type => type === dataType)
	) {
		const {end, start} = value as BetweenNumber;

		return isValid(end) && isValid(start);
	}

	switch (dataType) {
		case DataTypes.Boolean:
			return value === 'true' || value === 'false';
		case DataTypes.Duration:
		case DataTypes.Number:
			return isNumber(value);
		case DataTypes.Date:
		case DataTypes.String:
		default:
			return isValid(value);
	}
};
