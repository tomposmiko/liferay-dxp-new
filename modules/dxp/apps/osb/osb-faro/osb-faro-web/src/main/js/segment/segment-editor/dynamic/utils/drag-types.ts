export enum DragTypes {
	CriteriaGroup = 'criteria-group',
	CriteriaRow = 'criteria-row',
	Property = 'property'
}
