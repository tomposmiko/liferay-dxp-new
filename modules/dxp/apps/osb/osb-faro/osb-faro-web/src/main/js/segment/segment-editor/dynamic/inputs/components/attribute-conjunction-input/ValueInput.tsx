import DateInput from 'shared/components/DateInput';
import DateRangeInput, {DateRange} from 'shared/components/DateRangeInput';
import DurationInput from 'shared/components/DurationInput';
import Form from 'shared/components/form';
import getCN from 'classnames';
import Input from 'shared/components/Input';
import NumberInput from '../NumberInput';
import React from 'react';
import {BetweenNumber} from '../BetweenNumberInput';
import {BOOLEAN_OPTIONS} from 'event-analysis/utils/utils';
import {ClaySelectWithOption} from '@clayui/select';
import {createOption, validateAttributeValue} from './utils';
import {Criterion} from '../../../utils/types';
import {DataTypes} from 'event-analysis/utils/types';
import {FunctionalOperators} from '../../../utils/constants';
import {isValid} from '../../../utils/utils';

interface IValueInputProps {
	dataType: DataTypes;
	onChange: (params: {
		criterion?: Criterion;
		touched?: {attributeValue: boolean};
		valid?: {attributeValue: boolean};
	}) => void;
	operatorName: Criterion['operatorName'];
	touched: boolean;
	valid: boolean;
	value: string | number | BetweenNumber | DateRange;
}

const ValueInput: React.FC<IValueInputProps> = ({
	dataType,
	onChange,
	operatorName,
	touched,
	valid,
	value
}) => {
	const handleAttributeValueBlur = () => {
		onChange({
			touched: {attributeValue: true}
		});
	};

	switch (dataType) {
		case DataTypes.Boolean:
			return (
				<ClaySelectWithOption
					className='boolean-input'
					data-testid='attribute-value-boolean-input'
					onBlur={handleAttributeValueBlur}
					onChange={event => {
						const {value} = event.target;

						onChange({
							criterion: {value},
							touched: {attributeValue: true},
							valid: {
								attributeValue: validateAttributeValue(
									value,
									dataType
								)
							}
						});
					}}
					options={BOOLEAN_OPTIONS.map(option =>
						createOption(option, dataType)
					)}
				/>
			);
		case DataTypes.Date:
			if (operatorName === FunctionalOperators.Between) {
				return (
					<Form.GroupItem
						className={getCN({
							'has-error': !valid && touched
						})}
						shrink
					>
						<DateRangeInput
							onBlur={handleAttributeValueBlur}
							onChange={value => {
								onChange({
									criterion: {
										value
									},
									touched: {attributeValue: true},
									valid: {
										attributeValue:
											!!value.end && !!value.start
									}
								});
							}}
							overlayAlignment='rightCenter'
							value={value as DateRange}
						/>
					</Form.GroupItem>
				);
			}

			return (
				<Form.GroupItem
					className={getCN({
						'has-error': !valid && touched
					})}
					shrink
				>
					<DateInput
						onDateInputBlur={handleAttributeValueBlur}
						onDateInputChange={value => {
							onChange({
								criterion: {
									value
								},
								touched: {attributeValue: true},
								valid: {
									attributeValue: validateAttributeValue(
										value,
										dataType
									)
								}
							});
						}}
						overlayAlignment='rightCenter'
						value={value}
					/>
				</Form.GroupItem>
			);
		case DataTypes.Duration:
			return (
				<Form.GroupItem
					className={getCN({
						'has-error': !valid && touched
					})}
					shrink
				>
					<DurationInput
						onChange={value => {
							onChange({
								criterion: {
									value
								},
								touched: {attributeValue: true},
								valid: {attributeValue: isValid(value)}
							});
						}}
						value={value as string}
					/>
				</Form.GroupItem>
			);
		case DataTypes.Number:
			return (
				<NumberInput
					onChange={({touched, valid, value}) =>
						onChange({
							criterion: {value},
							touched: {attributeValue: touched},
							valid: {attributeValue: valid}
						})
					}
					operatorName={operatorName}
					touched={touched}
					valid={valid}
					value={value}
				/>
			);
		case DataTypes.String:
		default:
			return (
				<Form.GroupItem
					className={getCN({
						'has-error': !valid && touched
					})}
					shrink
				>
					<Input
						data-testid='attribute-value-string-input'
						onBlur={handleAttributeValueBlur}
						onChange={event => {
							const {value} = event.target;

							onChange({
								criterion: {value},
								touched: {attributeValue: true},
								valid: {
									attributeValue: validateAttributeValue(
										value,
										dataType
									)
								}
							});
						}}
						value={value}
					/>
				</Form.GroupItem>
			);
	}
};

export default ValueInput;
