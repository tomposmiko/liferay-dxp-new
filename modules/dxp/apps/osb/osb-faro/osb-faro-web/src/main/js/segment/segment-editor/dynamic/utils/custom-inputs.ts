import {createNewGroup} from './utils';
import {CustomValue} from 'shared/util/records';
import {fromJS, Map} from 'immutable';
import {isArray} from 'lodash';
import {TIME_PERIOD_OPTIONS} from './constants';

/**
 * Create the valueIMap for a custom input.
 * @param {Array} params - Array of root keys for the valueIMap.
 * @returns {CustomValue} The params array converted into a deeply immutable valueIMap.
 */
export const createCustomValueMap = (
	params: {key: string; value: any}[]
): CustomValue => {
	let valueIMap = Map();

	params.forEach(({key, value}) => {
		if (isArray(value)) {
			valueIMap = valueIMap.set(key, fromJS(createNewGroup(value)));
		} else {
			valueIMap = valueIMap.set(key, value);
		}
	});

	return new CustomValue(valueIMap);
};

/**
 * Get the filter Criterion Map at index.
 * @param {CustomValue} valueIMap - The Immutable Map representing the custom input value.
 * @param {number} index - The index of the Criterion Map.
 * @returns {Map} The Criterion Map.
 */
export const getFilterCriterionIMap = (valueIMap: CustomValue, index: number) =>
	valueIMap.getIn(['criterionGroup', 'items', index]);

/**
 * Get the index where the first matching property name is found in the criteria list.
 * This is useful if we don't know where the criterion is at in the list, but we know the property name.
 * @param {CustomValue} valueIMap - The Immutable Map representing the custom input value.
 * @returns {number} The index of the matching criterion or -1 if not found.
 */
export const getIndexFromPropertyName = (
	valueIMap: CustomValue,
	propertyName: string
): number =>
	valueIMap
		.getIn(['criterionGroup', 'items'])
		.findIndex(entry => entry.get('propertyName') === propertyName);

/**
 * Get the operator name from the criterion at the specified index in the valueIMap.
 * @param {CustomValue} valueIMap - The Immutable Map representing the custom input value.
 * @returns {string} The operator name of the criterion at the specified index.
 */
export const getOperator = (valueIMap: CustomValue, index: number): string =>
	getPropertyValue(valueIMap, 'operatorName', index);

/**
 * Get the time period value from the valueIMap.
 * @param {CustomValue} valueIMap - The Immutable Map representing the custom input value.
 * @returns {string} The time period value from the criterion with a 'completeDate' property name.
 */
export const getCompleteDate = (valueIMap: CustomValue): string => {
	const index = getIndexFromPropertyName(valueIMap, 'completeDate');

	return getPropertyValue(valueIMap, 'value', index);
};

/**
 * Get the value of the propertyName at the specified index.
 * @param {CustomValue} valueIMap - The Immutable Map representing the custom input value.
 * @param {string} propertyName - The propertyName string in the criterion to update.
 * @param {number} index - The index of the criterion in the items list.
 * @returns {*} The value of the criterion propertyName at the specified index in the items list.
 */
export const getPropertyValue = (
	valueIMap: CustomValue,
	propertyName: string,
	index: number
): any => valueIMap.getIn(['criterionGroup', 'items', index, propertyName]);

/**
 * Get the time period label from the time period value.
 * @param {string} timePeriod - The time period value.
 * @returns {string} The time period label that matches the timePeriodValue.
 */
export const getTimePeriodLabel = (value: string): string => {
	const timePeriod = TIME_PERIOD_OPTIONS.find(
		timePeriod => timePeriod.value === value
	);

	return timePeriod ? timePeriod.label : null;
};

/**
 * Remove entries in valueIMap by their index in the criteria list.
 * @param {CustomValue} valueIMap - The Immutable Map representing the custom input value.
 * @param {Array} indexArray - Array of indexes to remove from the criteria list in valueIMap.
 * @returns {CustomValue} The valueIMap with the items at the indexes in indexArray removed.
 */
export const removeItemsByIndex = (
	valueIMap: CustomValue,
	indexArray: number[]
): CustomValue =>
	valueIMap.updateIn(['criterionGroup', 'items'], iList =>
		iList.filterNot((_, i) => indexArray.includes(i))
	) as CustomValue;

/**
 * Set the operator name from the criterion at the specified index in the valueIMap.
 * @param {CustomValue} valueIMap - The Immutable Map representing the custom input value.
 * @param {number} index - The index of the criterion in the items list.
 * @param {*} value - The value to update in the valueIMap.
 * @returns {CustomValue} The updated valueIMap.
 */
export const setOperator = (
	valueIMap: CustomValue,
	index: number,
	value: any
): CustomValue => setPropertyValue(valueIMap, 'operatorName', index, value);

/**
 * Set the time period value in the valueIMap.
 * @param {CustomValue} valueIMap - The Immutable Map representing the custom input value.
 * @param {string} value - The value to update for the time period.
 * @returns {CustomValue} The updated valueIMap.
 */
export const setCompleteDate = (
	valueIMap: CustomValue,
	completeDate: string
): CustomValue => {
	const index = getIndexFromPropertyName(valueIMap, 'completeDate');

	return setPropertyValue(valueIMap, 'value', index, completeDate);
};

/**
 * Set the value of the propertyName at the specified index.
 * @param {CustomValue} valueIMap - The Immutable Map representing the custom input value.
 * @param {string} propertyName - The propertyName string in the criterion to update.
 * @param {number} index - The index of the criterion in the items list.
 * @param {*} value - The value to update in the valueIMap.
 * @returns {CustomValue} The updated valueIMap.
 */
export const setPropertyValue = (
	valueIMap: CustomValue,
	propertyName: string,
	index: number,
	value: any
): CustomValue =>
	valueIMap.setIn(
		['criterionGroup', 'items', index, propertyName],
		value
	) as CustomValue;
