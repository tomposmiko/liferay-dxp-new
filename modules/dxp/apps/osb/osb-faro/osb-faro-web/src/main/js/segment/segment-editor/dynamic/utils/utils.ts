import dateFns from 'date-fns';
import {
	Conjunctions,
	CustomFunctionOperators,
	isKnown,
	isUnknown,
	NotOperators,
	PropertyTypes,
	SUPPORTED_OPERATORS_MAP
} from './constants';
import {Criteria, Criterion, CriterionGroup, Operator} from './types';
import {Event} from 'event-analysis/utils/types';
import {every, isBoolean, isString, isUndefined} from 'lodash';
import {FieldContexts, FieldOwnerTypes} from 'shared/util/constants';
import {fromJS, Map} from 'immutable';
import {getUid} from 'metal';
import {
	INDIVIDUAL_PROPERTIES,
	ORGANIZATION_PROPERTIES,
	SESSION_PROPERTIES,
	WEB_BEHAVIORS
} from '../utils/properties';
import {Property} from 'shared/util/records';

const GROUP_ID_NAMESPACE = 'group_';
const ROW_ID_NAMESPACE = 'row_';

export const createInterestProperty = (name: string): Property =>
	new Property({
		entityName: Liferay.Language.get('individual'),
		label: name,
		name,
		propertyKey: 'interest',
		type: PropertyTypes.Interest
	});

/**
 * Creates a new group object with items.
 */
export const createNewGroup = (items: Criteria[]): CriterionGroup => ({
	conjunctionName: Conjunctions.And,
	criteriaGroupId: generateGroupId(),
	items
});

/**
 * Generates a unique group id.
 */
export const generateGroupId = (): string => `${GROUP_ID_NAMESPACE}${getUid()}`;

/**
 * Generates a unique row id.
 */
export const generateRowId = (): string => `${ROW_ID_NAMESPACE}${getUid()}`;

/**
 * Gets a list of group ids from a criteria object.
 * Used for disallowing groups to be moved into its own deeper nested groups.
 * Example of returned value: ['group_02', 'group_03']
 */
export const getChildGroupIds = (criteria: Criteria): string[] => {
	let childGroupIds = [];

	if (isCriterionGroup(criteria) && criteria.items.length) {
		childGroupIds = criteria.items.reduce(
			(groupIdList, item) =>
				isCriterionGroup(item)
					? [
							...groupIdList,
							item.criteriaGroupId,
							...getChildGroupIds(item)
					  ]
					: groupIdList,
			[]
		);
	}

	return childGroupIds;
};

/**
 * Gets the property name from the propertyLabel string .
 */
export const getPropertyNameFromRaw = (propertyLabel: string = ''): string => {
	const properties = propertyLabel.split('/');

	return properties.length > 1 ? properties[1] : properties[0];
};

export const getPropertyContextFromRaw = (
	propertyLabel: string = ''
): string => {
	const properties = propertyLabel.split('/');

	return properties.length > 1 ? properties[0] : null;
};

/**
 * Gets the list of operators for a supported type.
 * Used for displaying the operators available for each criteria row.
 */
export const getSupportedOperatorsFromType = (type: string = ''): Operator[] =>
	SUPPORTED_OPERATORS_MAP[type.toLowerCase()] || [];

/**
 * Checks if value is a CriterionGroup.
 */
export const isCriterionGroup = (
	value: CriterionGroup | Criterion
): value is CriterionGroup =>
	!!value && (value as CriterionGroup).items !== undefined;

/**
 * Checks if value is an ImmutableMap
 */
export const isMap = (
	value: Map<string, any> | object
): value is Map<string, any> => Map.isMap(value as Map<string, any>);

/**
 * Checks if value is either isKnown or isUnknown.
 */
export const isOfKnownType = (key: string): boolean =>
	[isKnown, isUnknown].includes(key);

/**
 * Converts an object of key value pairs to a form data object for passing
 * into a fetch body.
 */
export const objectToFormData = (dataObject: object): FormData => {
	const formData = new FormData();

	Object.keys(dataObject).forEach(key => {
		formData.set(key, dataObject[key]);
	});

	return formData;
};

/**
 * Parse an activityKey string into an object.
 */
export const parseActivityKey = (
	activityKey: string = ''
): {eventId: string; id: string; objectType: string} => {
	const [objectType, eventId, id] = activityKey.split('#');

	return {eventId, id, objectType};
};

/**
 * Returns a YYYY-MM-DD date
 * based on a JS Date object
 *
 * @export
 */
export const jsDatetoYYYYMMDD = (dateJsObject: Date): string => {
	const DATE_FORMAT = 'YYYY-MM-DD';
	return dateFns.format(dateJsObject, DATE_FORMAT);
};

/**
 * Finds the matching property based on its Criterion.
 */
export const findPropertyByCriterion = (
	criterion: Criterion,
	referencedPropertiesIMap: Map<string, Map<string, Property>>
): Property => {
	const {operatorName, propertyName, type, value} = criterion;

	if (
		[
			CustomFunctionOperators.ActivitiesFilterByCount,
			NotOperators.NotActivitiesFilterByCount
		].includes(operatorName)
	) {
		const {eventId = propertyName} = parseActivityKey(
			(value as Map<string, any>).getIn(
				['criterionGroup', 'items', 0, 'value'],
				''
			)
		);

		return WEB_BEHAVIORS.find(({name}) => name === eventId);
	} else if (
		[
			CustomFunctionOperators.EventsFilterByCount,
			NotOperators.NotEventsFilterByCount
		].includes(operatorName)
	) {
		const eventId = value.getIn(
			['criterionGroup', 'items', 0, 'value'],
			''
		);

		return referencedPropertiesIMap.getIn(['event', eventId]);
	} else if (
		[
			CustomFunctionOperators.AccountsFilter,
			NotOperators.NotAccountsFilter
		].includes(operatorName)
	) {
		return referencedPropertiesIMap.getIn(
			[
				'account',
				getPropertyContextFromRaw(propertyName),
				getPropertyNameFromRaw(propertyName)
			],
			''
		);
	} else if (
		[
			NotOperators.NotOrganizationsFilter,
			CustomFunctionOperators.OrganizationsFilter
		].includes(operatorName)
	) {
		if (getPropertyContextFromRaw(propertyName) !== FieldContexts.Custom) {
			return ORGANIZATION_PROPERTIES.find(
				({name}) => name === propertyName
			);
		}

		return referencedPropertiesIMap.getIn(
			[
				'organization',
				getPropertyContextFromRaw(propertyName),
				getPropertyNameFromRaw(propertyName)
			],
			''
		);
	} else if (
		[
			CustomFunctionOperators.SessionsFilter,
			NotOperators.NotSessionsFilter
		].includes(operatorName) ||
		type === PropertyTypes.SessionDateTime
	) {
		return SESSION_PROPERTIES.find(({name}) => name === propertyName);
	} else if (operatorName === CustomFunctionOperators.InterestsFilter) {
		return createInterestProperty(propertyName);
	} else if (INDIVIDUAL_PROPERTIES.find(({name}) => name === propertyName)) {
		return INDIVIDUAL_PROPERTIES.find(({name}) => name === propertyName);
	} else {
		return referencedPropertiesIMap.getIn(
			[
				'individual',
				getPropertyContextFromRaw(propertyName),
				getPropertyNameFromRaw(propertyName)
			],
			''
		);
	}
};

export const convertFieldMappingToAccountProperty = (
	fieldMapping:
		| Map<string, any>
		| {
				context: string;
				displayName: string;
				id: string;
				name: string;
				rawType: string;
		  }
): Property => {
	const context = isMap(fieldMapping)
		? fieldMapping.get('context')
		: fieldMapping.context;
	const displayName = isMap(fieldMapping)
		? fieldMapping.get('displayName')
		: fieldMapping.displayName;
	const id = isMap(fieldMapping) ? fieldMapping.get('id') : fieldMapping.id;
	const name = isMap(fieldMapping)
		? fieldMapping.get('name')
		: fieldMapping.name;
	const type = isMap(fieldMapping)
		? fieldMapping.get('rawType')
		: fieldMapping.rawType;

	const CUSTOM_REGEX = /custom-/;

	return new Property({
		entityName: Liferay.Language.get('account'),
		id,
		label: displayName || name,
		name: context ? `${context}/${name}/value` : name,
		propertyKey: FieldOwnerTypes.Account,
		type: CUSTOM_REGEX.test(type)
			? type.toLowerCase()
			: `account-${type.toLowerCase()}`
	});
};

export const convertFieldMappingToIndividualProperty = (
	fieldMapping:
		| Map<string, any>
		| {
				context: string;
				displayName: string;
				id: string;
				name: string;
				ownerType: string;
				rawType: string;
				type: string;
		  }
): Property => {
	const context = isMap(fieldMapping)
		? fieldMapping.get('context')
		: fieldMapping.context;
	const displayName = isMap(fieldMapping)
		? fieldMapping.get('displayName')
		: fieldMapping.displayName;
	const id = isMap(fieldMapping) ? fieldMapping.get('id') : fieldMapping.id;
	const name = isMap(fieldMapping)
		? fieldMapping.get('name')
		: fieldMapping.name;
	const type = isMap(fieldMapping)
		? fieldMapping.get('rawType')
		: fieldMapping.rawType;

	return new Property({
		entityName: Liferay.Language.get('individual'),
		id,
		label: displayName || name,
		name: context ? `${context}/${name}/value` : name,
		propertyKey: FieldOwnerTypes.Individual,
		type: type.toLowerCase()
	});
};

export const convertFieldMappingToOrganizationProperty = (
	fieldMapping:
		| Map<string, any>
		| {
				context: string;
				displayName: string;
				id: string;
				name: string;
				ownerType: string;
				rawType: string;
				type: string;
		  }
): Property => {
	const context = isMap(fieldMapping)
		? fieldMapping.get('context')
		: fieldMapping.context;
	const displayName = isMap(fieldMapping)
		? fieldMapping.get('displayName')
		: fieldMapping.displayName;
	const id = isMap(fieldMapping) ? fieldMapping.get('id') : fieldMapping.id;
	const name = isMap(fieldMapping)
		? fieldMapping.get('name')
		: fieldMapping.name;
	const type = isMap(fieldMapping)
		? fieldMapping.get('rawType')
		: fieldMapping.rawType;

	return new Property({
		entityName: Liferay.Language.get('organization'),
		id,
		label: displayName || name,
		name: context ? `${context}/${name}/value` : name,
		propertyKey: FieldOwnerTypes.Organization,
		type: `organization-${type.toLowerCase()}` as PropertyTypes
	});
};

export const convertEventToProperty = (
	eventDefinition: Map<string, any> | Event = Map()
): Map<string, Map<string, Property>> => {
	const displayName = isMap(eventDefinition)
		? eventDefinition.get('displayName')
		: eventDefinition.displayName;
	const id = isMap(eventDefinition)
		? eventDefinition.get('id')
		: eventDefinition.id;
	const name = isMap(eventDefinition)
		? eventDefinition.get('name')
		: eventDefinition.name;

	return new Property({
		entityName: Liferay.Language.get('event'),
		id,
		label: displayName || name,
		name: id,
		propertyKey: 'event',
		type: PropertyTypes.Event
	});
};

export const convertFieldMappingsToProperties = (
	fieldMappingsIMap: Map<
		string,
		Map<string, Map<string, Map<string, any>>>
	> = Map()
): Map<string, Map<string, Map<string, Property>>> =>
	fieldMappingsIMap.map((ownerTypeGroup, key) => {
		let conversionFn;

		if (key === FieldOwnerTypes.Account) {
			conversionFn = convertFieldMappingToAccountProperty;
		} else if (key === FieldOwnerTypes.Individual) {
			conversionFn = convertFieldMappingToIndividualProperty;
		} else if (key === FieldOwnerTypes.Organization) {
			conversionFn = convertFieldMappingToOrganizationProperty;
		}

		if (conversionFn) {
			return ownerTypeGroup.map(contextGroup =>
				contextGroup.reduce(
					(acc, fieldMappingIMap, key) =>
						acc.set(key, conversionFn(fieldMappingIMap)),
					Map()
				)
			);
		}
	}) as Map<string, Map<string, Map<string, Property>>>;

export const convertReferencedObjectsToProperties = (
	referencedObjectsIMap: Map<
		string,
		Map<string, Map<string, Map<string, any>>>
	> = Map()
): Map<string, Map<string, Map<string, Property> | Property>> => {
	const fieldMappingProperties = convertFieldMappingsToProperties(
		referencedObjectsIMap.get('fieldMappings')
	);

	const eventProperties = referencedObjectsIMap
		.get('event', Map())
		.map(convertEventToProperty);

	return fieldMappingProperties.merge(fromJS({event: eventProperties}));
};

/**
 * Check to see if the value is a valid input value.
 * The input value cannot be an empty string or undefined.
 * @returns {boolean}
 */
export const isValid = (value: any): boolean =>
	!(isUndefined(value) || (isString(value) && !value.length));

/**
 * Recursively check through all criterions and invalidates those
 * that do not have a matching property
 */
export const invalidateCriterionWithMissingProperty = (
	criteria: Criteria,
	referencedPropertiesIMap: Map<string, Property>
) => {
	if (isCriterionGroup(criteria)) {
		const {items} = criteria;

		if (items.length) {
			return {
				...criteria,
				items: items.map(criterion =>
					invalidateCriterionWithMissingProperty(
						criterion,
						referencedPropertiesIMap
					)
				)
			};
		}
	} else {
		if (findPropertyByCriterion(criteria, referencedPropertiesIMap)) {
			return criteria;
		}

		return {
			...criteria,
			valid: isBoolean(criteria.valid)
				? false
				: Object.keys(criteria.valid).reduce(
						(acc, key) => ({...acc, [key]: false}),
						{}
				  )
		};
	}
};

/**
 * Recursively check through all criteria to see if they're valid.
 */
export const validateSegmentInputs = (criteria: Criteria): boolean => {
	if (isCriterionGroup(criteria)) {
		const {items} = criteria;

		if (items.length) {
			return items.map(validateSegmentInputs).every(Boolean);
		}
	} else if (criteria) {
		if (isBoolean(criteria.valid)) {
			return criteria.valid;
		}

		return every(criteria.valid, Boolean);
	}
};
