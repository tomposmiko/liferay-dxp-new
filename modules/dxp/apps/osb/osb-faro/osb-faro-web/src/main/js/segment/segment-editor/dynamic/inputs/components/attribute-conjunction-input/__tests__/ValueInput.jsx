import React from 'react';
import ValueInput from '../ValueInput';
import {DataTypes} from 'event-analysis/utils/types';
import {
	FunctionalOperators,
	RelationalOperators
} from '../../../../utils/constants';
import {render} from '@testing-library/react';

jest.unmock('react-dom');

describe('ValueInput', () => {
	it('should render', () => {
		const {container} = render(
			<ValueInput
				dataType={DataTypes.Boolean}
				onChange={jest.fn()}
				operatorName={RelationalOperators.EQ}
				touched={false}
				valid
				value='true'
			/>
		);

		expect(container).toMatchSnapshot();
	});

	it.each`
		dataType             | operatorName                    | value                                       | testId
		${DataTypes.Boolean} | ${RelationalOperators.EQ}       | ${'true'}                                   | ${'attribute-value-boolean-input'}
		${DataTypes.Date}    | ${RelationalOperators.EQ}       | ${'2020-12-12'}                             | ${'date-input'}
		${DataTypes.Date}    | ${FunctionalOperators.Between}  | ${{end: '2020-12-12', start: '2020-12-01'}} | ${'date-range-input'}
		${DataTypes.Number}  | ${RelationalOperators.GT}       | ${1000}                                     | ${'number-input'}
		${DataTypes.Number}  | ${FunctionalOperators.Between}  | ${{end: 1000, start: 1}}                    | ${'between-number-end-input'}
		${DataTypes.String}  | ${FunctionalOperators.Contains} | ${'Stuff'}                                  | ${'attribute-value-string-input'}
	`(
		'should find $testId if dataType is $dataType and operatorName is $operatorName',
		({dataType, operatorName, testId, value}) => {
			const {queryByTestId} = render(
				<ValueInput
					dataType={dataType}
					onChange={jest.fn()}
					operatorName={operatorName}
					touched={false}
					valid
					value={value}
				/>
			);

			expect(queryByTestId(testId)).toBeTruthy();
		}
	);
});
