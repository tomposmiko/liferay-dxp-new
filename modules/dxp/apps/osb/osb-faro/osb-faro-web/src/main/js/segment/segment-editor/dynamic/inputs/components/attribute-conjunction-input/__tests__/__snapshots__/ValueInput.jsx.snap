// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValueInput should render 1`] = `
<div>
  <select
    class="form-control boolean-input"
    data-testid="attribute-value-boolean-input"
  >
    <option
      value="true"
    >
      True
    </option>
    <option
      value="false"
    >
      False
    </option>
  </select>
</div>
`;
